### DISCUSSION TITLE
Importing Data

### DISCUSSION THREAD

**Discussion by ZinnerC** [inventree/InvenTree#8629] [Q&A] [answered]
I'm currently setting up parts and their information within Inventree. For that purpose I first created all the parts with their basic information. Now I would like to add information step by step into Inventree.

But currently I experience the following situation with the old UI and its respective import functionality I only can import about 100-150 parts per import without getting a "internal server error". Not sure what it is but could be a timeout.

After that I tried to import some new information via the new GUI (PUI) and the respective new Data Importer. But somehow I do not manage to import anything.

Did I miss something? I know currently it is quite a generic question but would be interesting to hear from others how they experience the importing of additional data (actually updating existing parts)

**About my Inventree Installation**

- Digital Ocean Droplet with 2GB and regular CPU

# Version Information:
InvenTree-Version: 0.16.8
Django Version: 4.2.15

Database: postgresql
Debug-Mode: False
Deployed using Docker: False
Platform: Linux-5.4.0-200-generic-x86_64-with-glibc2.31
Installer: None

Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.1.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'InvenTreeCurrencyExchange', 'slug': 'inventreecurrencyexchange', 'version': '1.0.0'}, {'name': 'InvenTreeLabel', 'slug': 'inventreelabel', 'version': '1.1.0'}, {'name': 'InvenTreeLabelMachine', 'slug': 'inventreelabelmachine', 'version': '1.0.0'}, {'name': 'InvenTreeLabelSheet', 'slug': 'inventreelabelsheet', 'version': '1.0.0'}, {'name': 'DigiKeyPlugin', 'slug': 'digikeyplugin', 'version': '1.0.0'}, {'name': 'LCSCPlugin', 'slug': 'lcscplugin', 'version': '1.0.0'}, {'name': 'MouserPlugin', 'slug': 'mouserplugin', 'version': '1.0.0'}, {'name': 'TMEPlugin', 'slug': 'tmeplugin', 'version': '1.0.0'}]

**Comment by SchrodingersGat** [comment]
Hey @ZinnerC that import tool is soon to be removed (in our next major release when remove the old UI entirely).

This is because, the way that the `django-import-export` library imports data is extremely inefficient. If you have more than a handful of rows, it will timeout the server.

Some related issues with the same problem:

- https://github.com/inventree/InvenTree/issues/6376
- https://github.com/inventree/InvenTree/issues/5571

So, we have implemented a brand new importer - https://github.com/inventree/InvenTree/pull/6911

This is only available in the new user interface (you should be able to see it at `<host>/platform/settings/admin/import`. The new importer runs in the background and is much more efficient in any case.

**Comment by ZinnerC** [comment]
Hi @SchrodingersGat Thank you for your response yes I'm aware about the new importer.

I would like to show you my process so you could tell me where I'm doing something wrong. Or where I expect something wrong.

First I go to a part I would like to modify and downloads its data as a XLSX:
![Inventree_Import_Download](https://github.com/user-attachments/assets/a45cc9ea-5f76-45e9-a8fc-5a753a6383c6)
Then I check what ID the category I would like to add has (ID=1).
![Inventree_Import_Category](https://github.com/user-attachments/assets/049d83e1-1c9a-44b5-9d27-2e4a5ce3a797)
Then I edit the respective position in the XLSX (I upload the original download and the modified one.)
[InvenTree_Part_2024-12-06_Category.xlsx](https://github.com/user-attachments/files/18036042/InvenTree_Part_2024-12-06_Category.xlsx)
[InvenTree_Part_2024-12-06_Original.xlsx](https://github.com/user-attachments/files/18036041/InvenTree_Part_2024-12-06_Original.xlsx)

Then I go to the "Importer". There I add a new import and make the respective selections then I submit.
![Inventree_Import_ImportSettings](https://github.com/user-attachments/assets/0f53e5b9-2bc8-4925-a938-b5ecd3db7c5c)
- DataFile (modified XLSX)
- Model Type: Part since I downloaded that XLSX

There I do the mapping and only the the minimal (all others I ignore since I don't want to change and expect to have a more efficient import)
![Inventree_Import_Mapping_close](https://github.com/user-attachments/assets/fd0fa0d4-99fe-4ce8-a105-0f49d3ad10c5)
Complete Mapping:
![Inventree_Import_Mapping](https://github.com/user-attachments/assets/6ffd5a8e-5df4-4472-9f11-9f732b0c6c9e)

Then I check if it understood the input as expected (validate in step 4)
![Inventree_Import_Processing](https://github.com/user-attachments/assets/09016fcd-344e-4fb4-af4a-ee587d184615)
Then I select the one I would like to import and click on the green arrow.
![Inventree_Import_Processing_Action](https://github.com/user-attachments/assets/20e13313-1943-45cf-b751-3e46e5322df7)
At this very instance a popup appears on the lower right but quickly disappears again, telling something that something is going on.
But I don't reach step 5 where the import is completed.
Closing the dialog at step 4 and I see that the respective session is still avalable and not completed.
![Inventree_Import_AdminCenter](https://github.com/user-attachments/assets/494f4b77-a0e8-4c02-b781-d3adf20f4b69)
Quickly checking if the part has changed I see that it did not. However, this is not surprising since the import actually did not complete.

So this is how far I get I don't understand where somethings does not go as required.

**Reply by SchrodingersGat** [reply, marked as answer]
@ZinnerC the new importer only creates *new* records - does not allow for updating existing records in bulk

**Reply by ZinnerC** [reply]
Is it planned to actually also allow the import to update existing data? Would help in initiation.

**Reply by SchrodingersGat** [reply]
@ZinnerC TBH it had not been planned - but I have added a new issue here - https://github.com/inventree/InvenTree/issues/8640

If you want to help develop this, or support it with funding, that would be appreciated :)

**Reply by ZinnerC** [reply]
@SchrodingersGat  thank you for creating this issue. With respect of the development/funding I can check ;-) But I thank you for the for the support/information so far.

**Comment by SergeoLacruz** [comment]
When we started with InvenTree some years ago, we had the same problem importing ~2500 components with lots of properties and attachments from a huge excel table. We used pandas and the InvenTree python library to solve this. Using this method we were able to 
import everything including relations to suppliers and manufacturers, attachments, stock items, ...
Similar for BOM. CAD uses the python lib to create, or update a BOM without loosing build allocations and create a change list.

**Reply by ZinnerC** [reply]
Would you be able to share your script?

**Reply by SergeoLacruz** [reply]
Not really sorry. It just tinkered and not good enough for publishing because it was meant to used once. 
Parsing an excel table with pandas is easy. Finding all the information is very specific based on the structure of the table. 
Then I used the python functions to create parts, stockitems, and so on. Only creation, no modification. Just run it once category by category. When something went wrong, I deleted the parts and started a new run. Took some days.

**Reply by wolflu05** [reply]
Here are some links that may help you:
- https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html
- https://pandas.pydata.org/docs/user_guide/10min.html
- https://docs.inventree.org/en/latest/api/python/python/#item-attributes

**Reply by SergeoLacruz** [reply]
I wrote the major part of the sofa example during this work:

https://docs.inventree.org/en/latest/api/python/examples/

**Reply by ZinnerC** [reply]
Lovely. Thank you. I will have a try.
